### PR TITLE
Allow overriding a sensor's precision

### PR DESCRIPTION
**Pull request by emontnemery**
## Proposed change

Allow overriding a sensor's precision

Depends on core PR https://github.com/home-assistant/core/pull/86074

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by emontnemery** [comment]
Before a precision has been set, the text field renders like this, which is not so nice:
image

File: src/panels/config/entities/entity-registry-settings.ts [hunk 1]
```diff
@@ -459,6 +465,27 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
               </ha-select>
             `
           : ""}
+        ${domain === "sensor" &&
+        // Allow customizing the precision for a sensor with numerical device class,
+        // a unit of measurement or state class
+        ((this._deviceClass &&
+          this._deviceClass !== "date" &&
+          this._deviceClass !== "enum" &&
+          this._deviceClass !== "timestamp") ||
```
Comment (frenck) [inline src/panels/config/entities/entity-registry-settings.ts:474]: Could be written as `!["date", "enum", "timestamp"].includes(this._deviceClass)`

File: src/panels/config/entities/entity-registry-settings.ts [hunk 2]
```diff
@@ -459,6 +465,27 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
...
+          ? html`
+              <ha-textfield
+                .value=${this._precision}
+                .label=${this.hass.localize(
+                  "ui.dialogs.entity_registry.editor.precision"
+                )}
+                .invalid=${invalidDomainUpdate}
+                .disabled=${this._submitting}
+                @input=${this._precisionChanged}
+              ></ha-textfield>
```
Comment (frenck) [inline src/panels/config/entities/entity-registry-settings.ts:486]: Should this be numeric with a min/max?

```typescript
              <ha-textfield
                .inputMode="numeric"
                .min=0
                .max=10
                .value=${this._precision}
                .label=${this.hass.localize(
                  "ui.dialogs.entity_registry.editor.precision"
                )}
                .invalid=${invalidDomainUpdate}
                .disabled=${this._submitting}
                @input=${this._precisionChanged}
                type="number"
                autoValidate
              ></ha-textfield>
```

File: src/panels/config/entities/entity-registry-settings.ts [hunk 3]
```diff
@@ -1069,11 +1118,28 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
...
+          ...params.options,
+          unit_of_measurement: this._unit_of_measurement,
+        };
+      }
+      if (
+        domain === "sensor" &&
+        this.entry.options?.[domain]?.precision !== this._precision
+      ) {
+        params.options = { ...params.options, precision: this._precision };
+      }
```
Comment (piitaya) [inline src/panels/config/entities/entity-registry-settings.ts:1142]: Can we keep thing separated ? One block of code for number/sensor + unit and one block for sensor + precision.

File: src/translations/en.json [hunk 4]
```diff
@@ -905,6 +905,8 @@
           "entity_id": "Entity ID",
           "unit_of_measurement": "Unit of Measurement",
           "precipitation_unit": "Precipitation unit",
+          "precision": "Precision",
+          "precision_default": "Use default from integration",
```
Comment (piitaya) [inline src/translations/en.json:909]:
```suggestion
          "precision_default": "Use default",
```
Just "use default" ?

File: src/panels/config/entities/entity-registry-settings.ts [hunk 5]
```diff
@@ -459,6 +475,42 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
...
+        ((this._deviceClass &&
+          !["date", "enum", "timestamp"].includes(this._deviceClass)) ||
+          stateObj?.attributes.unit_of_measurement ||
+          stateObj?.attributes.state_class)
+          ? html`
+              <ha-select
+                .label=${this.hass.localize(
+                  "ui.dialogs.entity_registry.editor.precision"
+                )}
+                .value=${this._precision == null ? "default" : this._precision}
```
Comment (piitaya) [inline src/panels/config/entities/entity-registry-settings.ts:490]:
```suggestion
                .value=${this._precision == null ? "default" : this._precision.toString()}
```

File: src/panels/config/entities/entity-registry-settings.ts [hunk 6]
```diff
@@ -459,6 +475,42 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
...
+                @closed=${stopPropagation}
+              >
+                <mwc-list-item .value=${"default"}
+                  >${this.hass.localize(
+                    "ui.dialogs.entity_registry.editor.precision_default"
+                  )}</mwc-list-item
+                >
+                ${PRECISIONS.map(
+                  (precision) => html`
+                    <mwc-list-item .value=${precision}>
```
Comment (piitaya) [inline src/panels/config/entities/entity-registry-settings.ts:503]:
```suggestion
                    <mwc-list-item .value=${precision.toString()}>
```

File: src/panels/config/entities/entity-registry-settings.ts [hunk 7]
```diff
@@ -884,6 +936,11 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
     this._precipitation_unit = ev.target.value;
   }
 
+  private _precisionChanged(ev): void {
+    this._error = undefined;
+    this._precision = ev.target.value === "default" ? null : ev.target.value;
```
Comment (piitaya) [inline src/panels/config/entities/entity-registry-settings.ts:941]:
```suggestion
    this._precision = ev.target.value === "default" ? null : Number(ev.target.value);
```

**Review by piitaya — COMMENTED** [review]
I found the issue with the label.
the `ha-select` and `mwc-list-item` value must be "string" and not "number". I suggested some fixes in the code review.

**Comment by frenck** [comment]
Backend for this PR has been merged.

**Review by piitaya — APPROVED** [review]

**Pull request merged**